lls/cache.c

```diff
@@ -0,0 +1,428 @@
+			rte_mempool_put(cache->entry_pool, new_entry);
+		} else {
+			/* Add new entry (valid or invalid) to the cache. */
+			cache->data[ret] = new_entry;
+		}
+
+		if (get_lls_conf()->debug)
+			lls_cache_dump(cache);
+	}
+}
+
+static void
+__lls_cache_del(struct lls_cache *cache, const void *ip_key)
+{
+	int32_t ret = rte_hash_del_key(cache->hash, ip_key);
+	if (unlikely(ret == -ENOENT))
+		RTE_LOG(ERR, HASH, "No entry found, %s entry not deleted\n",
+			cache->name);
+	else if (unlikely(ret == -EINVAL))
+		RTE_LOG(ERR, HASH, "Invalid params, %s entry not deleted\n",
+			cache->name);
+	else {
+		if (unlikely(cache->data[ret]->refcnt > 0))
+			RTE_LOG(ERR, GATEKEEPER,
+				"Positive interest, %s entry not deleted\n",
+				cache->name);
+		else {
+			rte_mempool_put(cache->entry_pool, cache->data[ret]);
+			cache->data[ret] = NULL;
+			if (get_lls_conf()->debug)
+				lls_cache_dump(cache);
+		}
+	}
+}
+
+static void
+__lls_cache_refcnt(struct lls_config *lls_conf, struct lls_cache *cache,
+	const void *ip_key, struct lls_cache_entry *entry, enum lls_cache_op op)
+{
+	int ret = rte_hash_lookup(cache->hash, ip_key);
+	if (ret == -ENOENT) {
+		if (op == LLS_CACHE_INC) {
+			/* Add placeholder entry and try to resolve it. */
+			__lls_cache_add(cache, ip_key, entry);
+			cache->xmit_req(lls_conf->net->front.id, ip_key, NULL);
+			cache->xmit_req(lls_conf->net->back.id, ip_key, NULL);
+		} else
+			RTE_LOG(ERR, GATEKEEPER,
+				"lls: no such entry in %s cache to decrement refcnt\n",
+				cache->name);
+	} else if (unlikely(ret == -EINVAL)) {
+		rte_mempool_put(cache->entry_pool, entry);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: invalid parameters for %s cache in %s\n",
+			cache->name, __func__);
+	} else {
+		if (op == LLS_CACHE_INC) {
+			/* Placeholder entry was not needed. */
+			rte_mempool_put(cache->entry_pool, entry);
+			cache->data[ret]->refcnt++;
+		} else
+			cache->data[ret]->refcnt--;
+
+		if (get_lls_conf()->debug)
+			lls_cache_dump(cache);
+	}
+}
+
+unsigned int
+lls_cache_writer(struct lls_config *lls_conf)
+{
+	struct lls_cache_req *reqs[LLS_CACHE_BURST_SIZE];
+	unsigned int count = mb_dequeue_burst(&lls_conf->mb,
+		(void **)reqs, LLS_CACHE_BURST_SIZE);
+	unsigned int i;
+
+	if (likely(count == 0))
+		return 0;
+
+	for (i = 0; i < count; i++) {
+		switch (reqs[i]->op) {
+		case LLS_CACHE_ADD:
+			__lls_cache_add(reqs[i]->cache, reqs[i]->ip_key,
+				reqs[i]->entry);
+			break;
+		case LLS_CACHE_INC:
+			/* FALLTHROUGH */
+		case LLS_CACHE_DEC:
+			__lls_cache_refcnt(lls_conf, reqs[i]->cache,
+				reqs[i]->ip_key, reqs[i]->entry, reqs[i]->op);
+			break;
+		default:
+			RTE_LOG(ERR, GATEKEEPER,
+				"lls: unrecognized %s cache opcode (%d)\n",
+				reqs[i]->cache->name, reqs[i]->op);
+			break;
+		}
+		mb_free_entry(&lls_conf->mb, reqs[i]);
+	}
+
+	return count;
+}
+
+struct lls_cache_entry *
+lls_cache_get(struct lls_cache *cache, const void *ip_key)
+{
+	int ret;
+
+	if (unlikely(cache == NULL)) {
```

Comment:
There's nothing that prevents or discourages other blocks from calling functions like `arp_put()`, `arp_hold()`, etc. until after the LLS block is initialized (and these functions try to use the cache). I mean, blocks *should* wait to do this until after the network starts and all blocks are running, but there's nothing enforcing it.